src/drivers/gpio/bcm283x/bcm283x_gpio.c

```diff
@@ -0,0 +1,332 @@
+
+#define REG_BITS 32
+
+#define PBASE                       OPTION_GET(NUMBER,base_addr)   
+#define BCM283X_PINS_NUMBER         OPTION_GET(NUMBER,gpio_pins)
+#define BCM283X_GPIO_PORTS_COUNT    OPTION_GET(NUMBER,gpio_ports)
+#define BCM283X_GPIO_JTAG           OPTION_GET(BOOLEAN,gpio_jtag)
+
+#define BCM283X_PULL_PORTA_DEFLT_FLOAT    OPTION_GET(NUMBER,gpu_pullup_porta_float)
+#define BCM283X_PULL_PORTA_DEFLT_DOWN     OPTION_GET(NUMBER,gpu_pullup_porta_down)
+#define BCM283X_PULL_PORTA_DEFLT_UP       OPTION_GET(NUMBER,gpu_pullup_porta_up)
+
+#define BCM283X_PULL_PORTB_DEFLT_FLOAT    OPTION_GET(NUMBER,gpu_pullup_portb_float)
+#define BCM283X_PULL_PORTB_DEFLT_DOWN     OPTION_GET(NUMBER,gpu_pullup_portb_down)
+#define BCM283X_PULL_PORTB_DEFLT_UP       OPTION_GET(NUMBER,gpu_pullup_portb_up)
+
+#define PORT_PIN_TO_PINNUMBER(pt, pn)   ( ( ((pt) & 0x3 ) * REG_BITS ) + ((pn) % REG_BITS )) 
+#define GPIO_ALTERNATE_INVERT(af) ( ((af) >> 17) & 0x07 )
+#define BCM283X_GPIO_CHIP_ID    0
+
+#define IRQ_NR_PORTA  49
+#define IRQ_NR_PORTB  50
+#define IRQ_NR_PORTC  51
+#define IRQ_NR_ALL    52
+
+typedef enum _GpioPullUpDn {
+    GPUDNone = 0,
+    GPUDDown = 1,
+    GPUDUp  = 2
+} GpioPullUpDn;
+
+#define GPIO_PULL_NONE_UP_DN_STATES 3
+
+struct GpioPinData {
+    /* 0 (reserved) - Pin output set 2 - GPIO pins 64-96 (reserved on Pi Zero)
+     * 1 (data[0]) - Pin output set 0 - GPIO pins 0-31
+     * 2 (data[1])- Pin output set 1 - GPIO pins 32-53 (54-63 not connected/reserved)
+     */
+    volatile uint32_t reserved;
+    volatile uint32_t data[BCM283X_GPIO_PORTS_COUNT];  
+};
+
+struct GpioRegs {
+    volatile uint32_t func_select[3 * BCM283X_GPIO_PORTS_COUNT];
+    struct GpioPinData output_set;              // Set as output
+    struct GpioPinData output_clear;            // Set as input, is default
+    struct GpioPinData level;                   // Read for value of pin
+    struct GpioPinData ev_detect_status;        // Edge event detected
+    struct GpioPinData re_detect_enable;        // Synchoronous rising edge event
+    struct GpioPinData fe_detect_enable;        // Synchoronous falling edge event
+    struct GpioPinData hi_detect_enable;        // Hi state detected
+    struct GpioPinData lo_detect_enable;        // Lo state detected
+    struct GpioPinData re_async_detect_enable;  // Enable async rising detection in ev_detect_status
+                                                //     (no sampling by clock cycle for edge detection)
+    struct GpioPinData fe_async_detect_enable;  // Enable async falling detection in ev_detect_status
+    volatile uint32_t reserved;
+    volatile uint32_t pupd_enable;              // See func bcm283x_gpio_pin_pullupdown() below
+    volatile uint32_t pupd_enable_clocks[BCM283X_GPIO_PORTS_COUNT];
+};
+
+#define REGS_GPIO ((struct GpioRegs *)(PBASE))
+
+//// Globals
+//
+uint32_t _pinMaskPortA[GPIO_PULL_NONE_UP_DN_STATES];
+uint32_t _pinMaskPortB[GPIO_PULL_NONE_UP_DN_STATES];
```

Comment:
static ?

Reply:
Yes, if there is another way, I'd like to know.  The issue is that setting of the pull up/down/float state must be done across all pins of a given port at once.  You can't just change one pin at a time.  Also, the setting is persistent over power on/off cycle.  Also, you can't read the setting (that is the big difference from other systems?).  So, I have global defaults set as per data sheet, then with every setup call, modify those defaults and always with every call, define all pull up/down/floats.  The globals retain the configuration between each gpio setup call.